pyomo/contrib/mindtpy/feasibility_pump.py

```diff
@@ -0,0 +1,330 @@
+
+    # if OA-like or feas_pump converged, update Upper bound,
+    # add no_good cuts and increasing objective cuts (feas_pump)
+    if feas_pump_converged(solve_data, config):
+        copy_var_list_values(solve_data.mip.MindtPy_utils.variable_list,
+                             solve_data.working_model.MindtPy_utils.variable_list,
+                             config)
+        fix_nlp, fix_nlp_results = solve_NLP_subproblem(
+            solve_data, config)
+        assert fix_nlp_results.solver.termination_condition is tc.optimal, 'Feasibility pump fix-nlp subproblem not optimal'
+        copy_var_list_values(fix_nlp.MindtPy_utils.variable_list,
+                             solve_data.working_model.MindtPy_utils.variable_list,
+                             config)
+        if main_objective.sense == minimize:
+            solve_data.UB = min(
+                main_objective.expr(),
+                solve_data.UB)
+            solve_data.solution_improved = \
+                solve_data.UB < solve_data.UB_progress[-1]
+            solve_data.UB_progress.append(solve_data.UB)
+
+            if solve_data.solution_improved and config.strategy == 'feas_pump':
+                solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.\
+                    increasing_objective_cut.set_value(
+                        expr=solve_data.mip.MindtPy_utils.objective_value
+                        <= solve_data.UB - config.feas_pump_delta*min(1e-4, abs(solve_data.UB)))
+        else:
+            solve_data.LB = max(
+                main_objective.expr(),
+                solve_data.LB)
+            solve_data.solution_improved = \
+                solve_data.LB > solve_data.LB_progress[-1]
+            solve_data.LB_progress.append(solve_data.LB)
+
+            if solve_data.solution_improved and config.strategy == 'feas_pump':
+                solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.\
+                    increasing_objective_cut.set_value(
+                        expr=solve_data.mip.MindtPy_utils.objective_value
+                        >= solve_data.LB + config.feas_pump_delta*min(1e-4, abs(solve_data.LB)))
+
+        if config.add_no_good_cuts or config.strategy is 'feas_pump':
+            config.logger.info('Creating no-good cut')
+            add_nogood_cuts(solve_data.mip, config)
+    else:
+        solve_data.solution_improved = False
+
+    config.logger.info(
+        'NLP {}: OBJ: {}  LB: {}  UB: {}'
+        .format(solve_data.nlp_iter,
+                main_objective.expr(),
+                solve_data.LB, solve_data.UB))
+
+    if solve_data.solution_improved:
+        solve_data.best_solution_found = solve_data.working_model.clone()
+        assert is_feasible(solve_data.best_solution_found, config), \
+            "Best found model infeasible! There might be a problem with the precisions - the feaspump seems to have converged (error**2 <= integer_tolerance). " \
```

Comment:
Best found "solution" I guess. And I don't understand when can this ever happen.